File: backend/app/models/organization.py
"""
Organization model for multi-tenancy.

Each dental clinic is an organization with its own users, patients, and data.
"""

from datetime import datetime
from typing import Optional
from uuid import uuid4

from sqlalchemy import Boolean, Column, DateTime, Enum, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class SubscriptionTier(str, enum.Enum):
    """Subscription tiers for pricing."""

    BASIC = "basic"  # Free tier - Core 4 agents only
    PROFESSIONAL = "professional"  # $1,500/month - + CFO & Operations agents
    ENTERPRISE = "enterprise"  # $4,500/month - All 7 executive agents


class Organization(Base):
    """Organization model for multi-tenancy."""

    __tablename__ = "organizations"

    # Primary key
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)

    # Basic info
    name = Column(String(255), nullable=False, index=True)
    slug = Column(String(255), unique=True, nullable=False, index=True)
    description = Column(Text, nullable=True)

    # Contact
    email = Column(String(255), nullable=False)
    phone = Column(String(20), nullable=True)
    address = Column(Text, nullable=True)

    # Subscription
    subscription_tier = Column(
        Enum(SubscriptionTier), nullable=False, default=SubscriptionTier.BASIC
    )
    subscription_status = Column(String(50), nullable=False, default="active")
    subscription_start_date = Column(DateTime, nullable=True)
    subscription_end_date = Column(DateTime, nullable=True)

    # Billing
    stripe_customer_id = Column(String(255), nullable=True, unique=True)
    stripe_subscription_id = Column(String(255), nullable=True, unique=True)

    # Odoo integration
    odoo_db_name = Column(String(255), nullable=True, unique=True)
    odoo_api_key = Column(String(255), nullable=True)

    # Status
    is_active = Column(Boolean, default=True, nullable=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # Soft delete
    deleted_at = Column(DateTime, nullable=True)

    # Relationships
    users = relationship("User", back_populates="organization")

    def __repr__(self) -> str:
        return f"<Organization {self.name}>"

File: frontend/src/types/organization.ts
/**
 * Organization Types
 * 
 * Multi-tenancy support for DentaFlow.
 * Each dental clinic is an organization.
 * 
 * Backend: app/models/organization.py
 */

/**
 * Subscription tier for pricing
 */
export enum SubscriptionTier {
  Basic = 'basic',
  Professional = 'professional',
  Enterprise = 'enterprise',
}

/**
 * Subscription status
 */
export enum SubscriptionStatus {
  Active = 'active',
  Trial = 'trial',
  Suspended = 'suspended',
  Cancelled = 'cancelled',
}

/**
 * Organization role in membership
 */
export enum OrganizationRole {
  Owner = 'owner',
  Manager = 'manager',
  Staff = 'staff',
  Patient = 'patient',
}

/**
 * Organization entity (dental clinic)
 */
export interface Organization {
  id: string;
  name: string;
  slug: string;
  description?: string;
  
  // Contact
  email: string;
  phone?: string;
  address?: string;
  
  // Subscription
  subscription_tier: SubscriptionTier;
  subscription_status: SubscriptionStatus;
  subscription_start_date?: string;
  subscription_end_date?: string;
  
  // Billing
  stripe_customer_id?: string;
  stripe_subscription_id?: string;
  
  // Odoo Integration
  odoo_db_name?: string;
  
  // Status
  is_active: boolean;
  
  // Timestamps
  created_at: string;
  updated_at: string;
}

/**
 * User's membership in an organization
 */
export interface OrganizationMembership {
  id: string;
  user_id: string;
  organization_id: string;
  
  // Organization details (denormalized for convenience)
  organization_name: string;
  organization_slug: string;
  
  // Roles
  organization_role: OrganizationRole;
  functional_role?: string;  // dentist, hygienist, receptionist, etc.
  
  // Odoo link
  odoo_partner_id?: number;
  
  // Status
  is_active: boolean;
  joined_at: string;
}

/**
 * Current organization context
 */
export interface OrganizationContext {
  organization: Organization;
  membership: OrganizationMembership;
}

/**
 * Organization registration request
 */
export interface OrganizationRegisterRequest {
  // Clinic info
  clinic_name: string;
  clinic_email: string;
  clinic_phone?: string;
  clinic_address?: string;
  
  // Owner info
  owner_full_name: string;
  owner_email: string;
  owner_password: string;
  owner_phone?: string;
}

/**
 * Organization registration response
 */
export interface OrganizationRegisterResponse {
  organization_id: string;
  organization_name: string;
  organization_slug: string;
  owner_id: string;
  owner_email: string;
  access_token: string;
  token_type: string;
  message: string;
}

/**
 * Agent availability by subscription tier
 */
export const AGENTS_BY_TIER: Record<SubscriptionTier, string[]> = {
  [SubscriptionTier.Basic]: ['alex', 'sarah', 'harper', 'sophia'],
  [SubscriptionTier.Professional]: ['alex', 'sarah', 'harper', 'sophia', 'marcus'],
  [SubscriptionTier.Enterprise]: ['alex', 'sarah', 'harper', 'sophia', 'marcus', 'cfo', 'cto'],
};

/**
 * Check if agent is available in subscription tier
 * 
 * @param agentName - Name of the agent
 * @param tier - Subscription tier
 * @returns true if agent is available in tier
 * 
 * @example
 * ```typescript
 * const canUseAgent = isAgentAvailable('marcus', SubscriptionTier.Professional);
 * // true
 * ```
 */
export function isAgentAvailable(agentName: string, tier: SubscriptionTier): boolean {
  return AGENTS_BY_TIER[tier].includes(agentName.toLowerCase());
}

/**
 * Get subscription tier display name
 * 
 * @param tier - Subscription tier
 * @returns Display name with pricing
 * 
 * @example
 * ```typescript
 * const displayName = getSubscriptionTierName(SubscriptionTier.Professional);
 * // "Professional ($1,500/month)"
 * ```
 */
export function getSubscriptionTierName(tier: SubscriptionTier): string {
  switch (tier) {
    case SubscriptionTier.Basic:
      return 'Basic (Free)';
    case SubscriptionTier.Professional:
      return 'Professional ($1,500/month)';
    case SubscriptionTier.Enterprise:
      return 'Enterprise ($4,500/month)';
  }
}

/**
 * Get subscription status display name
 * 
 * @param status - Subscription status
 * @returns Display name
 * 
 * @example
 * ```typescript
 * const displayName = getSubscriptionStatusName(SubscriptionStatus.Trial);
 * // "Trial"
 * ```
 */
export function getSubscriptionStatusName(status: SubscriptionStatus): string {
  switch (status) {
    case SubscriptionStatus.Active:
      return 'Active';
    case SubscriptionStatus.Trial:
      return 'Trial';
    case SubscriptionStatus.Suspended:
      return 'Suspended';
    case SubscriptionStatus.Cancelled:
      return 'Cancelled';
  }
}
